File: identity-android/src/main/java/com/android/identity/android/direct_access/DirectAccessSocketTransport.java
/*
 * Copyright 2023 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.identity.android.direct_access;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetAddress;
import java.net.Socket;

public class DirectAccessSocketTransport implements DirectAccessTransport {

  private static final int PORT = 8080;
  private static final String IPADDR = "192.168.9.112";
  private static final int MAX_RECV_BUFFER_SIZE = 700;
  private Socket mSocket;
  private boolean socketStatus;

  @Override
  public void init() throws IOException {

  }

  @Override
  public void openConnection() throws IOException {
    if (!isConnected()) {
      InetAddress serverAddress = InetAddress.getByName(IPADDR);
      mSocket = new Socket(serverAddress, PORT);
      socketStatus = true;
    }
  }

  @Override
  public byte[] sendData(byte[] inData) throws IOException {
    int count = 1;
    while (!socketStatus && count++ < 5) {
      try {
        Thread.sleep(1000);
        System.out.println("SocketTransport Trying to open socket connection... count: " + count);
        openConnection();
      } catch (InterruptedException e) {
        e.printStackTrace();
      }
    }

    if (count >= 5) {
      System.out.println("SocketTransport Failed to open socket connection");
      return null;
    }

    // Prepend the input length to the inputData before sending.
    byte[] length = new byte[2];
    length[0] = (byte) ((inData.length >> 8) & 0xFF);
    length[1] = (byte) (inData.length & 0xFF);

    try {
      ByteArrayOutputStream bs = new ByteArrayOutputStream();
      bs.write(length);
      bs.write(inData);
      OutputStream outputStream = mSocket.getOutputStream();
      outputStream.write(bs.toByteArray());
      outputStream.flush();
    } catch (IOException e) {
      System.out.println(
          "SocketTransport Failed to send data over socket. Error: " + e.getMessage());
      return null;
    }

    return readData();
  }

  @Override
  public void closeConnection() throws IOException {
    if (mSocket != null) {
      mSocket.close();
      mSocket = null;
    }
    socketStatus = false;
  }

  @Override
  public boolean isConnected() throws IOException {
    return socketStatus;
  }

  @Override
  public int getMaxTransceiveLength() {
    return MAX_RECV_BUFFER_SIZE;
  }

  @Override
  public void unInit() throws IOException {

  }

  private byte[] readData() {
    byte[] buffer = new byte[MAX_RECV_BUFFER_SIZE];
    short expectedResponseLen = 0;
    int totalBytesRead = 0;
    ByteArrayOutputStream bs = new ByteArrayOutputStream();

    try {
      do {
        short offset = 0;
        InputStream inputStream = mSocket.getInputStream();
        int numBytes = inputStream.read(buffer, 0, MAX_RECV_BUFFER_SIZE);
        if (numBytes < 0) {
          System.out.println("SocketTransport Failed to read data from socket.");
          return null;
        }
        totalBytesRead += numBytes;
        if (expectedResponseLen == 0) {
          expectedResponseLen |= (buffer[1] & 0xFF);
          expectedResponseLen |= ((buffer[0] << 8) & 0xFF00);
          expectedResponseLen += 2;
          numBytes -= 2;
          offset = 2;
        }
        bs.write(buffer, offset, numBytes);
      } while (totalBytesRead < expectedResponseLen);
      return bs.toByteArray();
    } catch (IOException e) {
      e.printStackTrace();
      return null;
    }
  }
}

File: identity/src/main/java/com/android/identity/direct_access/DirectAccessSocketTransport.kt
package com.android.identity.direct_access

import java.io.ByteArrayOutputStream
import java.io.IOException
import java.net.InetAddress
import java.net.Socket

class DirectAccessSocketTransport : DirectAccessTransport {
    private val PORT = 8080
    private val IPADDR = "192.168.9.112"
    private val MAX_RECV_BUFFER_SIZE = 700
    private var mSocket: Socket? = null
    private var socketStatus = false

    @Throws(IOException::class)
    override fun init() {
    }

    @Throws(IOException::class)
    override fun openConnection() {
        if (!isConnected) {
            val serverAddress = InetAddress.getByName(IPADDR)
            mSocket = Socket(serverAddress, PORT)
            socketStatus = true
        }
    }

    @Throws(IOException::class)
    override fun sendData(inData: ByteArray): ByteArray {
        var count = 1
        while (!socketStatus && count++ < 5) {
            try {
                Thread.sleep(1000)
                println("SocketTransport Trying to open socket connection... count: $count")
                openConnection()
            } catch (e: InterruptedException) {
                e.printStackTrace()
            }
        }
        if (count >= 5) {
            throw IOException("SocketTransport Failed to open socket connection")
        }

        // Prepend the input length to the inputData before sending.
        val length = ByteArray(2)
        length[0] = (inData.size shr 8 and 0xFF).toByte()
        length[1] = (inData.size and 0xFF).toByte()
        try {
            val bs = ByteArrayOutputStream()
            bs.write(length)
            bs.write(inData)
            val outputStream = mSocket!!.getOutputStream()
            outputStream.write(bs.toByteArray())
            outputStream.flush()
        } catch (e: IOException) {
            throw IOException("SocketTransport Failed to send data over socket. Error: " + e.message)
        }
        return readData()
    }

    @Throws(IOException::class)
    override fun closeConnection() {
        if (mSocket != null) {
            mSocket!!.close()
            mSocket = null
        }
        socketStatus = false
    }

    override val isConnected: Boolean
        get() = socketStatus

    override val maxTransceiveLength: Int
        get() = MAX_RECV_BUFFER_SIZE

    @Throws(IOException::class)
    override fun unInit() {
    }

    private fun readData(): ByteArray {
        val buffer = ByteArray(MAX_RECV_BUFFER_SIZE)
        var expectedResponseLen = 0
        var totalBytesRead = 0
        val bs = ByteArrayOutputStream()
        do {
            var offset: Short = 0
            val inputStream = mSocket!!.getInputStream()
            var numBytes = inputStream.read(buffer, 0, MAX_RECV_BUFFER_SIZE)
            if (numBytes < 0) {
                throw IOException("SocketTransport Failed to read data from socket.")
            }
            totalBytesRead += numBytes
            if (expectedResponseLen == 0) {
                expectedResponseLen = (buffer[1].toInt() and 0xFF)
                expectedResponseLen = (expectedResponseLen or (buffer[0].toInt() shl 8 and 0xFF00))
                expectedResponseLen += 2
                numBytes -= 2
                offset = 2
            }
            bs.write(buffer, offset.toInt(), numBytes)
        } while (totalBytesRead < expectedResponseLen)
        return bs.toByteArray()
    }
}